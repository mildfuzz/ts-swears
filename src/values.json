{
  "ddl": {
    "page": {
      "nbs_environment": null,
      "nbs_server": null,
      "pageInfo": {
        "pageName": null,
        "nbs_page_load_time": null,
        "nbs_page_responsive_state": null,
        "nbs_page_responsive_orientation": null
      },
      "category": {
        "primaryCategory": "over payment application",
        "nbs_sub_category_1": null,
        "nbs_sub_category_2": null,
        "nbs_sub_category_3": null
      }
    },

    "nbs_message": [],
    "nbs_user_input": [],
    "nbs_element_interaction": [],
    "nbs_app_mortgage": {},

    "nbs_user": {
      "nbs_user_application_route": null
    }
  },

  "regionProperties": {
    "view": ["page"],
    "action": ["nbs_user_input", "nbs_element_interaction", "nbs_message"],
    "context": ["nbs_user", "nbs_app_mortgage"]
  },

  "itemsKeyMap": {
    "nbs_message": {
      "category": null,
      "text": null,
      "cause": null
    },

    "nbs_user_input": {
      "name": "nbs_input_name",
      "status": "nbs_input_validation_status",
      "message": "nbs_input_validation_message"
    },

    "nbs_element_interaction": {
      "type": "nbs_interaction_type",
      "label": "nbs_interaction_label"
    }
  },

  "page": {
    "PREFIX": "nbs-aws",
    "TYPE": "application",
    "APP_NAME": "OverPayment",
    "INITIAL": "initial load",

    "types": {
      "LOADER": "loader"
    },
    "orientations": {
      "PORTRAIT": "portrait",
      "LANDSCAPE": "landscape",
      "SQUARE": "square"
    },
    "sizes": {
      "SMALL": "small",
      "MEDIUM": "medium",
      "LARGE": "large"
    }
  },

  "valueMaps": {
    "environments": {
      "home-mumo-dev": "dev",
      "home-mumo-prd": "live",
      "home-afford-dev-rtl": "dev",
      "home-afford-sit-rtl": "sit"
    },

    "routes": {
      "referrers": {
        "://(www.)?nationwide.co.uk": "nationwide.co.uk",
        "://onlinebanking.nationwide.co.uk": "internet bank",
        "^$": "external"
      },
      "urlParams": {
        "JId": "banking app"
      }
    },

    "inputValidity": {
      "true": "passed",
      "false": "failed",
      "null": "none"
    }
  },

  "types": {
    "action": {
      "SUBMIT": "submit",
      "INPUT": "user_input_complete",
      "DEFAULT": "misc_interaction"
    },

    "interaction": {
      "?LINK": "external link",
      "SUBMIT": "submit",
      "REVEAL": "reveal",
      "HELP": "help icon",
      "PRINT": "print",
      "PDF": "save pdf",
      "CALL": "click to call",
      "BACK": "back",
      "MISC": "misc"
    },

    "message": {
      "USER": "user error",
      "BUSINESS": "business error",
      "SYSTEM": "system error",
      "NOTIFICATION": "notification"
    }
  }
}
