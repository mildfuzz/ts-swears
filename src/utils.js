import _get from 'lodash/get'
import _map from 'lodash/map'
import _join from 'lodash/join'
import _trim from 'lodash/trim'
import _take from 'lodash/take'
import _find from 'lodash/find'
import _round from 'lodash/round'
import _concat from 'lodash/concat'
import _reduce from 'lodash/reduce'
import _compact from 'lodash/compact'
import _findKey from 'lodash/findKey'
import _toLower from 'lodash/toLower'
import _includes from 'lodash/includes'
import _isObject from 'lodash/isObject'
import _isUndefined from 'lodash/isUndefined'

import values from './values.json'

/**
 * Page name formatting
 * Tagging: Spec - 4.5.1.3
 */
export const getDdlText = (text) => _trim(_toLower(text || ''))

/**
 * Returns page name string from name sections array
 * Tagging: Spec - 4.5.1.3
 *
 * @param {array} sections
 * @returns {string} lowercase concatenated name sections
 */
export const toDdlName = (sections) =>
  _join(_compact(_map(sections, getDdlText)), ':')

/**
 * Returns page name sections using global and page specific sections
 * Tagging: Spec - 4.5.1.3
 *
 * @param {string|array} desc
 * @returns {array} page name sections
 */
export const getPageNameSections = (desc) => {
  const {PREFIX, TYPE, APP_NAME, APP_SUBNAME} = values.page
  return _concat(_compact([PREFIX, TYPE, APP_NAME, APP_SUBNAME]), desc)
}

export const getPageName = (desc) => toDdlName(getPageNameSections(desc))

export const getPageCategories = (desc) => {
  const pageNameSections = getPageNameSections(desc)

  return {
    nbs_sub_category_1: toDdlName(_take(pageNameSections, 2)),
    nbs_sub_category_2: toDdlName(_take(pageNameSections, 3)),
    nbs_sub_category_3: toDdlName(_take(pageNameSections, 4)),
  }
}

/**
 * Returns the region for ddl property
 * @param {string} property
 * @returns {string} region
 */
export const getPropertyRegion = (property) =>
  _findKey(
    values.regionProperties,
    (properties, region) => _includes(properties, property) && region,
  )

/**
 * Checks if property is context specific
 * @param {string} property
 * @return {bool}
 */
export const isContextProperty = (property) =>
  getPropertyRegion(property) === 'context'

/**
 * Checks if property is action specific
 * @param {string} property
 * @return {bool}
 */
export const isActionProperty = (property) =>
  getPropertyRegion(property) === 'action'

/**
 * Returns the key for a DDL item, if item or key not found returns key provided
 * @param {string} property
 * @param {string} item
 * @returns {string} DDL item property key
 */
export const getItemPropertyKey = (item, property) =>
  _get(values.itemsKeyMap, [item, property], property) || property

/**
 * Load time measuring
 * Tagging: Spec - 5.1.6.1, Req. - 1.2
 */
export const pageLoadTimes = {
  [getPageName(values.page.INITIAL)]: global.initialLoadStart,
}

const timeNow = () =>
  global?.performance?.now ? global?.performance?.now() : Date?.now()

export const loadStart = (pageName) => {
  const loadTime = timeNow()
  pageLoadTimes[pageName] = loadTime
  return loadTime
}

export const loadEnd = (pageName) => {
  const loadTime = _round(Math.abs(timeNow() - pageLoadTimes[pageName]))
  pageLoadTimes[pageName] = loadTime
  return loadTime
}

export const getEntryRoute = () => {
  const {
    referrer,
    location: {search},
  } = global.document
  const {referrers, urlParams} = values.valueMaps.routes
  const defaultRoute = referrers['^$']
  let entryRoute = _find(referrers, (_, pattern) =>
    new RegExp(pattern).test(referrer),
  )

  if (!entryRoute || entryRoute === defaultRoute) {
    entryRoute =
      _find(urlParams, (_, pattern) =>
        new RegExp(`[\\?&]${pattern}=`, 'i').test(search),
      ) || entryRoute
  }

  return entryRoute || defaultRoute
}

/**
 * Technical service data
 * Tagging: Spec - 5.1.6.1, Req. - 1.2
 */
export const getServiceData = (env) => ({
  nbs_environment: env,
  nbs_server: global.location.origin,
})

/**
 * Device data
 * Tagging: Spec - 5.1.6.1, Req. - 1.3
 */
export const getDeviceData = ({width, height}) => {
  const {orientations, sizes} = values.page

  let orientation
  let state

  switch (width % height) {
    case 0:
      orientation = orientations.SQUARE
      break
    case width:
      orientation = orientations.PORTRAIT
      break
    default:
      orientation = orientations.LANDSCAPE
      break
  }

  if (width <= 480) state = sizes.SMALL
  else if (width <= 768) state = sizes.MEDIUM
  else state = sizes.LARGE

  return {
    nbs_page_responsive_orientation: orientation,
    nbs_page_responsive_state: state,
  }
}

/**
 * converts formik error data structure to analytics message structure
 */
export const formatFormikErrors = (formikErrors) => {
  const getMappedErrors = (err, curPath) =>
    _reduce(
      err,
      (acc, val, key) => {
        const path = !_isUndefined(curPath) ? `${curPath}.${key}` : key

        if (_isObject(val)) {
          return {...acc, ...getMappedErrors(val, path)}
        }

        return {
          ...acc,
          [path]: val,
        }
      },
      {},
    )

  const errors = getMappedErrors(formikErrors)

  return _map(errors, (message, cause) => ({message, cause}))
}
