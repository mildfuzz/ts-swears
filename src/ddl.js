import _get from 'lodash/get'
import _set from 'lodash/set'
import _pick from 'lodash/pick'
import _mapKeys from 'lodash/mapKeys'
import _cloneDeep from 'lodash/cloneDeep'
import _defaultsDeep from 'lodash/defaultsDeep'
import _uniqBy from 'lodash/uniqBy'

import * as utils from './utils'
import values from './values.json'

const {valueMaps, regionProperties} = values
const getInitialViewData = () =>
  _defaultsDeep(
    {},
    _pick(values.ddl, regionProperties.view, regionProperties.action),
  )

const contextData = _defaultsDeep(
  {},
  _pick(values.ddl, regionProperties.context),
)
let viewData = getInitialViewData()

/**
 * Internal API
 */
const ddl = {
  /**
   * Get property data source
   * @param {string} property
   */
  getSource: (property) => {
    if (utils.isContextProperty(property)) return contextData
    if (utils.isActionProperty(property) || property === 'page')
      return global.digitalData
    return viewData
  },

  /**
   * Get data from ddl
   * @param {string} property
   */
  get: (property) => _get(ddl.getSource(property), property),

  /**
   * Create new ddl
   * @param {object} data
   */
  create: (data) => {
    global.digitalData = _cloneDeep(
      _defaultsDeep(data, _cloneDeep(contextData), viewData),
    )
    viewData = getInitialViewData()
  },

  /**
   * Merge contextual data with ddl
   */
  refresh: () => {
    global.digitalData = _defaultsDeep(
      {},
      _cloneDeep(contextData),
      global.digitalData,
    )
  },

  /**
   * Update existing ddl data
   * @param {string} property
   * @param {*} data
   */
  update: (property, newData, applyDefaults = true) => {
    const source = ddl.getSource(property)
    const data = applyDefaults
      ? _defaultsDeep(_cloneDeep(newData), ddl.get(property))
      : _cloneDeep(newData)
    _set(source, property, data)
  },

  /**
   * Add ddl item to a ddl array - new items are always at head
   * @param {string} itemProperty
   * @param {*} item
   */
  add: (itemProperty, item, applyDefaults = true) => {
    const ddlItems = ddl.get(itemProperty) || []
    const ddlItem = _mapKeys(item, (_, property) =>
      utils.getItemPropertyKey(itemProperty, property),
    )

    ddl.update(itemProperty, [ddlItem, ...ddlItems], applyDefaults)
  },

  /**
   * Add ddl item to a ddl array - new items are always at head
   * @param {string} itemProperty
   * @param {*} item
   */
  replace: (itemProperty, item, applyDefaults = true) => {
    const ddlItems = ddl.get(itemProperty) || []
    const ddlItem = _mapKeys(item, (_, property) =>
      utils.getItemPropertyKey(itemProperty, property),
    )

    const messages = _uniqBy([ddlItem, ...ddlItems], 'cause')

    ddl.update(itemProperty, messages, applyDefaults)
  },
}

export default {
  /**
   * spec - 5.1.6.1
   * @param {string|array} desc
   */
  ...ddl,

  setPage: (desc, isProd = false) => {
    const {getPageName} = utils
    const pageName = getPageName(desc)
    const category = ''

    ddl.create({
      page: {
        pageInfo: {
          pageName,
          nbs_page_load_time: utils.loadEnd(pageName),
          ...utils.getDeviceData({
            width: global.innerWidth,
            height: global.innerHeight,
          }),
        },
        category,
        ...utils.getServiceData(isProd ? 'live' : 'dev'),
      },
    })
  },

  updatePageName: (desc) => {
    const {getPageName} = utils
    const pageName = getPageName(desc)
    const category = ''

    ddl.update('page', {
      pageInfo: {
        pageName,
      },
      category,
    })

    ddl.refresh()
  },

  setUserApplicationRoute: ({applicationRoute}) => {
    ddl.update('nbs_user', {
      nbs_user_application_route: utils.getDdlText(applicationRoute),
    })
  },

  /**
   * Spec: 5.1.6.5
   */
  addMessage: (category, text, cause) => {
    ddl.add('nbs_message', {
      category: utils.getDdlText(category),
      text: utils.getDdlText(text),
      cause: utils.getDdlText(cause),
    })
  },

  replaceMessage: (category, text, cause) => {
    ddl.replace('nbs_message', {
      category: utils.getDdlText(category),
      text: utils.getDdlText(text),
      cause: utils.getDdlText(cause),
    })
  },

  /**
   * Spec: 5.2.1.2
   */
  addUserInput: (name, message, isValid = null, applyDefaults = true) => {
    ddl.add(
      'nbs_user_input',
      {
        name: utils.getDdlText(name),
        message: utils.getDdlText(message),
        status: valueMaps.inputValidity[isValid],
      },
      applyDefaults,
    )
  },

  /**
   * Spec: 5.2.X.3
   */
  addInteraction: (type, label) => {
    ddl.add('nbs_element_interaction', {
      type: utils.getDdlText(type),
      label: utils.getDdlText(label),
    })
  },

  updateMortgageObject: (analyticsValues) =>
    ddl.update('nbs_app_mortgage', analyticsValues),
  /**
   * Additional utilities
   */
  refresh: () => ddl.refresh(),
}
