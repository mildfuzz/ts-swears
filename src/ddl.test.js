import _pick from 'lodash/pick'
import _forEach from 'lodash/forEach'

import * as utils from './utils'
import values from './values.json'
import ddl from './ddl'

describe('@analytics > ddl', () => {
  jest.spyOn(global.Date, 'now').mockReturnValue(0)
  jest.spyOn(global, 'wa_view').mockImplementation()

  const tagPage = (desc) => {
    global.Date.now.mockReturnValue(0)
    utils.loadStart(utils.getPageName(desc))
    global.Date.now.mockReturnValue(5000) // 5 second page load
    ddl.setPage(desc)
  }

  beforeEach(() => jest.clearAllMocks())

  describe('getSource()', () => {
    test('getSource returns context data if called with a context property', () => {
      expect(ddl.getSource('nbs_user')).toMatchSnapshot()
    })

    test('getSource returns global digital if called with an action property', () => {
      expect(ddl.getSource('nbs_user_input')).toMatchSnapshot()
    })

    test('getSource returns global digital if called with the page property', () => {
      expect(ddl.getSource('page')).toMatchSnapshot()
    })

    test('getSource returns initial view data otherwise', () => {
      expect(ddl.getSource('other_property')).toMatchSnapshot()
    })
  })

  describe('updatePageName()', () => {
    test('update page name directly in digitalData object', () => {
      const getData = () =>
        _pick(global.digitalData, ['page.pageInfo.pageName', 'page.category'])
      const result = {before: null, after: null}

      tagPage('foo')
      result.before = getData()

      ddl.updatePageName(['bar 1', 'bar 2'])
      result.after = getData()

      expect(result).toMatchSnapshot()
    })
  })

  describe('setUserApplicationRoute()', () => {
    test('updates ddl with user route', () => {
      ddl.setUserApplicationRoute({
        applicationRoute: 'banking app',
      })

      tagPage()

      expect(
        _pick(global.digitalData, ['nbs_user_application_route']),
      ).toMatchSnapshot()
    })
  })

  describe('addMessage()', () => {
    test('adds contextual message item', () => {
      _forEach(values.types.message, (messageType) => {
        ddl.addMessage(
          messageType,
          `message for ${messageType}`,
          `cause for ${messageType}`,
        )
      })
      expect(_pick(global.digitalData, 'nbs_message')).toMatchSnapshot()
    })
  })

  describe('replceMessage()', () => {
    test('adds contextual message item', () => {
      _forEach(values.types.message, (messageType) => {
        ddl.replaceMessage(
          messageType,
          `message for ${messageType}`,
          `cause for ${messageType}`,
        )
      })

      expect(_pick(global.digitalData, 'nbs_message')).toMatchSnapshot()
    })
  })

  describe('addUserInput()', () => {
    test('adds contextual user input item with optional message, validation status, and apply defaults flag', () => {
      tagPage()
      ddl.addUserInput('foo input 1')
      ddl.addUserInput('foo input 2', 'foo input 2 success')
      ddl.addUserInput('foo input 3', null, true)
      ddl.addUserInput('foo input 4', 'foo input 4 passed', true)
      ddl.addUserInput('foo input 5', 'foo input 5 error', false)
      ddl.addUserInput('foo input 6', null, true, false)
      ddl.addUserInput('foo input 7', null, true, true)

      expect(_pick(global.digitalData, 'nbs_user_input')).toMatchSnapshot()
    })
  })

  describe('addInteraction()', () => {
    test('adds contextual interaction item', () => {
      tagPage()
      _forEach(values.types.interaction, (interactionType) => {
        ddl.addInteraction(interactionType, `label for ${interactionType}`)
      })

      expect(
        _pick(global.digitalData, 'nbs_element_interaction'),
      ).toMatchSnapshot()
    })
  })

  describe('updateMortgageObject()', () => {
    test('creates an array of names & values', () => {
      const analyticsValues = {
        customerType: 'New mortgage customer',
        transactionType: 'First time buyer',
      }
      ddl.updateMortgageObject(analyticsValues)

      expect(_pick(global.digitalData, 'page')).toEqual({
        page: {
          category: '',
          nbs_environment: 'dev',
          nbs_server: 'http://localhost',
          pageInfo: {
            nbs_page_load_time: 0,
            nbs_page_responsive_orientation: 'landscape',
            nbs_page_responsive_state: 'large',
            pageName: 'nbs-aws:application:overpayment',
          },
        },
      })
    })
  })

  describe('refresh()', () => {
    test('merges contextual data with ddl', () => {
      ddl.addUserInput('name', 'message', true)
      ddl.refresh()

      expect(_pick(global.digitalData, 'nbs_user_input')).toMatchSnapshot()
    })
  })
})
