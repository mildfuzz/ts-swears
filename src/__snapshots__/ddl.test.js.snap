// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@analytics > ddl addInteraction() adds contextual interaction item 1`] = `
Object {
  "nbs_element_interaction": Array [
    Object {
      "nbs_interaction_label": "label for misc",
      "nbs_interaction_type": "misc",
    },
    Object {
      "nbs_interaction_label": "label for back",
      "nbs_interaction_type": "back",
    },
    Object {
      "nbs_interaction_label": "label for click to call",
      "nbs_interaction_type": "click to call",
    },
    Object {
      "nbs_interaction_label": "label for save pdf",
      "nbs_interaction_type": "save pdf",
    },
    Object {
      "nbs_interaction_label": "label for print",
      "nbs_interaction_type": "print",
    },
    Object {
      "nbs_interaction_label": "label for help icon",
      "nbs_interaction_type": "help icon",
    },
    Object {
      "nbs_interaction_label": "label for reveal",
      "nbs_interaction_type": "reveal",
    },
    Object {
      "nbs_interaction_label": "label for submit",
      "nbs_interaction_type": "submit",
    },
    Object {
      "nbs_interaction_label": "label for external link",
      "nbs_interaction_type": "external link",
    },
  ],
}
`;

exports[`@analytics > ddl addMessage() adds contextual message item 1`] = `
Object {
  "nbs_message": Array [
    Object {
      "category": "notification",
      "cause": "cause for notification",
      "text": "message for notification",
    },
    Object {
      "category": "system error",
      "cause": "cause for system error",
      "text": "message for system error",
    },
    Object {
      "category": "business error",
      "cause": "cause for business error",
      "text": "message for business error",
    },
    Object {
      "category": "user error",
      "cause": "cause for user error",
      "text": "message for user error",
    },
  ],
}
`;

exports[`@analytics > ddl addUserInput() adds contextual user input item with optional message, validation status, and apply defaults flag 1`] = `
Object {
  "nbs_user_input": Array [
    Object {
      "nbs_input_name": "foo input 7",
      "nbs_input_validation_message": "",
      "nbs_input_validation_status": "passed",
    },
    Object {
      "nbs_input_name": "foo input 6",
      "nbs_input_validation_message": "",
      "nbs_input_validation_status": "passed",
    },
    Object {
      "nbs_input_name": "foo input 5",
      "nbs_input_validation_message": "foo input 5 error",
      "nbs_input_validation_status": "failed",
    },
    Object {
      "nbs_input_name": "foo input 4",
      "nbs_input_validation_message": "foo input 4 passed",
      "nbs_input_validation_status": "passed",
    },
    Object {
      "nbs_input_name": "foo input 3",
      "nbs_input_validation_message": "",
      "nbs_input_validation_status": "passed",
    },
    Object {
      "nbs_input_name": "foo input 2",
      "nbs_input_validation_message": "foo input 2 success",
      "nbs_input_validation_status": "none",
    },
    Object {
      "nbs_input_name": "foo input 1",
      "nbs_input_validation_message": "",
      "nbs_input_validation_status": "none",
    },
  ],
}
`;

exports[`@analytics > ddl getSource() getSource returns context data if called with a context property 1`] = `
Object {
  "nbs_app_mortgage": Object {},
  "nbs_user": Object {
    "nbs_user_application_route": null,
  },
}
`;

exports[`@analytics > ddl getSource() getSource returns global digital if called with an action property 1`] = `
Object {
  "page": Object {
    "pageInfo": Object {
      "pageName": "",
    },
  },
}
`;

exports[`@analytics > ddl getSource() getSource returns global digital if called with the page property 1`] = `
Object {
  "page": Object {
    "pageInfo": Object {
      "pageName": "",
    },
  },
}
`;

exports[`@analytics > ddl getSource() getSource returns initial view data otherwise 1`] = `
Object {
  "nbs_element_interaction": Array [],
  "nbs_message": Array [],
  "nbs_user_input": Array [],
  "page": Object {
    "category": Object {
      "nbs_sub_category_1": null,
      "nbs_sub_category_2": null,
      "nbs_sub_category_3": null,
      "primaryCategory": "over payment application",
    },
    "nbs_environment": null,
    "nbs_server": null,
    "pageInfo": Object {
      "nbs_page_load_time": null,
      "nbs_page_responsive_orientation": null,
      "nbs_page_responsive_state": null,
      "pageName": null,
    },
  },
}
`;

exports[`@analytics > ddl refresh() merges contextual data with ddl 1`] = `
Object {
  "nbs_user_input": Array [
    Object {
      "nbs_input_name": "name",
      "nbs_input_validation_message": "message",
      "nbs_input_validation_status": "passed",
    },
  ],
}
`;

exports[`@analytics > ddl replceMessage() adds contextual message item 1`] = `
Object {
  "nbs_message": Array [
    Object {
      "category": "notification",
      "cause": "cause for notification",
      "text": "message for notification",
    },
    Object {
      "category": "system error",
      "cause": "cause for system error",
      "text": "message for system error",
    },
    Object {
      "category": "business error",
      "cause": "cause for business error",
      "text": "message for business error",
    },
    Object {
      "category": "user error",
      "cause": "cause for user error",
      "text": "message for user error",
    },
  ],
}
`;

exports[`@analytics > ddl setUserApplicationRoute() updates ddl with user route 1`] = `Object {}`;

exports[`@analytics > ddl updatePageName() update page name directly in digitalData object 1`] = `
Object {
  "after": Object {
    "page": Object {
      "category": "",
      "pageInfo": Object {
        "pageName": "nbs-aws:application:overpayment:bar 1:bar 2",
      },
    },
  },
  "before": Object {
    "page": Object {
      "category": "",
      "pageInfo": Object {
        "pageName": "nbs-aws:application:overpayment:foo",
      },
    },
  },
}
`;
